lib/queue.js

```diff
@@ -1,256 +1,497 @@
-      this.jobs[message.id].options.retries -= 1;
+      this.jobs.get(message.id).options.retries -= 1;
     }
 
-    this.jobs[message.id].emit(message.event, message.data);
+    this.jobs.get(message.id).emit(message.event, message.data);
 
     if (message.event === 'succeeded' || message.event === 'failed') {
-      delete this.jobs[message.id];
+      this.jobs.delete(message.id);
     }
   }
 };
 
-Queue.prototype.close = function (cb) {
-  cb = cb || helpers.defaultCb;
-  this.paused = true;
+Queue.prototype.isRunning = function () {
+  return !this.paused;
+};
 
-  /* istanbul ignore next */
-  var closeTimeout = setTimeout(function () {
-    return cb(Error('Timed out closing redis connections'));
-  }, 5000);
+Queue.prototype.ready = function (cb) {
+  if (cb) this._ready.then(() => cb(null), cb);
+
+  return this._ready;
+};
 
-  var clients = [this.client];
-  if (this.settings.isWorker) {
-    clients.push(this.bclient);
+Queue.prototype._commandable = function (requireBlocking) {
+  if (requireBlocking ? this.paused : this._isClosed) {
+    return Promise.reject(new Error('closed'));
   }
 
-  if (this.settings.getEvents) {
-    clients.push(this.eclient);
+  if (this._isReady) {
+    return Promise.resolve(requireBlocking ? this.bclient : this.client);
   }
 
-  var handleEnd = barrier(clients.length, function () {
-    clearTimeout(closeTimeout);
-    return cb(null);
-  });
+  return this._ready.then(() => this._commandable(requireBlocking));
+};
+
+Queue.prototype.close = function (timeout, cb) {
+  if (typeof timeout === 'function') {
+    cb = timeout;
+    timeout = defaults['#close'].timeout;
+  } else if (!Number.isSafeInteger(timeout) || timeout <= 0) {
+    timeout = defaults['#close'].timeout;
+  }
+
+  if (this.paused) {
+    return this._closed;
+  }
+
+  this.paused = true;
+
+  if (this.checkTimer) {
+    clearTimeout(this.checkTimer);
+    this.checkTimer = null;
+  }
+
+  this._delayedTimer.stop();
+
+  const cleanup = () => {
+    this._isClosed = true;
 
-  clients.forEach(function (client) {
-    client.end();
-    client.stream.on('close', handleEnd);
+    const clients = [this.client];
+    if (this.settings.getEvents) {
+      clients.push(this.eclient);
+    }
+
+    return Promise.all(clients.map((client) => {
+      const promise = helpers.deferred();
+      client.quit(promise.defer());
+      return promise;
+    }));
+  };
+
+  const closed = helpers.withTimeout(this._ready.then(() => {
+    // Stop the blocking connection, ensures that we don't accept additional
+    // jobs while waiting for the ongoing jobs to terminate.
+    if (this.settings.isWorker) {
+      this.bclient.end(true);
+    }
+
+    // Wait for all the jobs to complete. Ignore job errors during shutdown.
+    const waitJobs = Array.from(this.activeJobs);
+    return Promise.all(waitJobs.map((promise) => promise.catch(() => {})));
+  }), timeout).then(() => {
+    return cleanup().then(() => undefined);
```

Comment:
I almost want to say `=> void 0` or `=> {}` here, but that's paranoid-level defensive programming. Reads better like this.